### PR TITLE
API: Extend FileIO in optional interfaces

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR extends `FileIO` in its optional interfaces. This way, we we can access all methods available in `FileIO` if we have a reference to the optional interface. Otherwise, we have to cast back.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
@danielcweeks @rdblue @flyrain @szehon-ho @stevenzwu @jackye1995, could you take a look?

File: api/src/main/java/org/apache/iceberg/io/CredentialSupplier.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
  * the configured credential as a string, and use the credential for file access via other IO
  * libraries.
  */
-public interface CredentialSupplier {
+public interface CredentialSupplier extends FileIO {
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/io/CredentialSupplier.java:29]: All these interfaces assume they are mixed-in by `FileIO` instances (see docs).

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/CredentialSupplier.java:29]: This is not quite correct. `CredentialSupplier` may extend other interfaces and not just `FileIO`.

**Comment by rdblue** [comment]
I don't think we should do this for `CredentialSupplier`, but the other two look fine.

File: api/src/main/java/org/apache/iceberg/io/CredentialSupplier.java (hunk 1 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/io/CredentialSupplier.java:29]: You are right. I misinterpreted the doc.

**Comment by stevenzwu** [comment]
@aokolnychyi this doesn't seem to be a problem in current code. I assume you ran into this need of casting in some new code you are working on?

I found the interface name of `SupportsPrefixOperations`  less intuitive for an interface extending from `FileIO`. Supports just meant some extra capability.  can't connect `Supports...` as an extended interface of `FileIO`.

Another downside is that `S3FileIO` would need to implement both `SupportsBulkOperations` and `SupportsPrefixOperations` with large overlap of base `FileIO` interface. not a problem for compiler. just thought a little odd.

**Comment by aokolnychyi** [comment]
@stevenzwu, this becomes an issue when we need both the optional functionality as well the core `FileIO` methods. I don't think these `SupportsXXX` can be used without `FileIO` as they are tightly coupled with it. It's the same pattern that Spark, for example, uses in its public APIs. These `SupportsXXX` simply add optional functionality to `FileIO`.

**Comment by stevenzwu** [comment]
agree that `SupportsXXX` can't be used without `FileIO`, as `SupportsXXX` declares additional capability for `FileIO`. 
```
public class S3FileIO
    implements FileIO, SupportsBulkOperations, SupportsPrefixOperations
```

I am trying to understand the concrete issue (type cast problem) that this tries to help with. Are we trying to type cast a `FileIO` object to `SupportsXXX` interface to check the capability? Can we use `instanceOf` to check the capability?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @aokolnychyi!

**Pull request merged**